.github/workflows/contracts-build-test.yml

```diff
@@ -0,0 +1,135 @@
+              - 'contracts/**'
+              - '.github/workflows/contracts-build-test.yml'
+
+  build:
+    name: Build and test contracts
+    needs: [ changes ]
+    # Run if changes are detected OR we're manually dispatching (for testing/release)
+    if: ${{ needs.changes.outputs.ton_changes == 'true' || github.event_name == 'workflow_dispatch' }}
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+    steps:
+      - name: Check out code
+        uses: actions/checkout@0ad4b8fadaa221de15dcec353f45205ec38ea70b # v4.1.4
+        with:
+          fetch-depth: 0
+
+      - name: Get Long and Short SHAs
+        id: get_sha
+        run: |
+          set -euo pipefail
+          TARGET_SHA="${{ github.event_name == 'workflow_dispatch' && github.event.inputs.sha != '' && github.event.inputs.sha || github.sha }}"
+          FULL_SHA="$(git rev-parse "${TARGET_SHA}")"
+          {
+            echo "short_sha=${FULL_SHA:0:12}"
+            echo "full_sha=${FULL_SHA}"
+          } >> "$GITHUB_OUTPUT"
+
+      - name: Install Nix
+        uses: cachix/install-nix-action@02a151ada4993995686f9ed4f1be7cfbb229e56f # v31
+        with:
+          nix_path: nixpkgs=channel:nixos-unstable
+
+      - name: Build (yarn)
+        run: |
+          set -euo pipefail
+          pushd contracts
+          nix develop .#contracts -c yarn && yarn build
+          popd
+
+      - name: Run tests
+        run: |
+          set -euo pipefail
+          pushd contracts
+          nix develop .#contracts -c  yarn test
+          popd
+
+
+      - name: Package build outputs
+        id: package
+        run: |
+          set -euo pipefail
+          mkdir -p dist
+          TAR_NAME="ton-contracts-build-${{ steps.get_sha.outputs.short_sha }}.tar.gz"
+          tar -C contracts \
+              --exclude="build/examples.*" \
+              -czf "dist/${TAR_NAME}" build
+          echo "tar_path=dist/${TAR_NAME}" >> "$GITHUB_OUTPUT"
+
+      - name: Upload artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: ton-contracts-build-${{ steps.get_sha.outputs.short_sha }}
+          path: ${{ steps.package.outputs.tar_path }}
+          if-no-files-found: error
+          retention-days: 14
+
+      # Create a release on push to main OR when manually requested
+      - name: Publish Release
+        if: ${{ (github.event_name == 'push' && github.ref == 'refs/heads/main') || (github.event_name == 'workflow_dispatch' && github.event.inputs.create_release == 'true') }}
+        uses: softprops/action-gh-release@c95fe1489396fe8a9eb87c0abf8aa5b2ef267fda # v2.2.1
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          tag_name: ton-contracts-build-${{ steps.get_sha.outputs.short_sha }}
+          target_commitish: ${{ steps.get_sha.outputs.full_sha }}
+          name: TON Contracts Build (${{ steps.get_sha.outputs.short_sha }})
+          draft: false
+          prerelease: false
+          files: |
```

Comment:
Why do we actually need this?

Can't we just:

- During active development, run `TON_CONTRACTS_DIR=$(nix build 'git+https://github.com/smartcontractkit/chainlink-ton#contracts' --print-out-paths)` will get you the latest from anywhere (e.g., core repo, cld repo)
- Once we have a stable version, we make a release, and pin it where we need it `TON_CONTRACTS_DIR=$(nix build 'git+https://github.com/smartcontractkit/chainlink-ton?ref=v0.1.1#contracts' --print-out-paths)`

Not sure why we need custom releases on every commit for this.